Product reviews

to6ko12 opened the pull request:
[sc-24913]

to6ko12 committed:
Add docker-compose configuration to run project locally

to6ko12 committed:
Add new config value in form for product review import + styling for checks if reviews module is installed and enabled.

to6ko12 committed:
Send order data with correct keys to the dashboard.

to6ko12 committed:
Rename option for consistency

to6ko12 committed:
Add API endpoint to sync product reviews

to6ko12 committed:
Fix error response for api endpoint + some formatting

to6ko12 committed:
Add functionality to sync product reviews from Dashboard to Prestashop

TsankoBg commented on common/src/Module.php:
```diff
@@ -6,6 +6,7 @@
 use Db;
 use Image;
 use Link;
+use Matrix\Exception;
```
What is this Exception class?

TsankoBg commented on common/src/Module.php:
```diff
@@ -81,9 +83,57 @@ public function install() {
         Configuration::updateGlobalValue($this->getConfigName('INVITE'), '');
         Configuration::updateGlobalValue($this->getConfigName('JAVASCRIPT'), '1');
 
+        $sync_url = _PS_BASE_URL_ . __PS_BASE_URI__ . sprintf('module/%s/sync', $this->getName());
```
Don't hardcode the URL, check documentation https://devdocs.prestashop-project.org/8/modules/concepts/controllers/front-controllers/#accessing-a-module-front-controller

TsankoBg commented on common/src/Module.php:
```diff
@@ -81,9 +83,57 @@ public function install() {
         return true;
     }
 
+    private function sendSyncUrl(string $sync_url, int $shop_id): void {
```
No need to create local variables `$sync_url` and `$shop_id` and then pass them to `sendSyncUrl`.
Both fields can be retrieved.

TsankoBg commented on common/src/Module.php:
```diff
@@ -81,9 +83,57 @@ public function install() {
     }
 
+    private function sendSyncUrl(string $sync_url, int $shop_id): void {
+        if (empty(Configuration::get(strtoupper($this->getName()) . '_SYNC_PROD_REVIEWS'))) {
```
Why don't you use `getConfigName`?

TsankoBg replied:
No need for `empty`, it can be just `!`
`Configuration::get` returns either the value or `NULL`

TsankoBg commented on common/src/Module.php:
```diff
@@ -81,9 +83,57 @@ public function install() {
+        $url = sprintf(self::SYNC_URL, $this->getDashboardDomain());
+        $data = [
+            'webshop_id' => $shop_id,
+            'api_key' => Configuration::get(strtoupper($this->getName()) . '_API_KEY'),
```
Why don't you use `getConfigName`?

TsankoBg commented on common/src/Module.php:
```diff
@@ -422,6 +450,22 @@ private function sendInvites($ps_shop_id) {
+            $line['sku'] = $line['product_reference'];
+            $line['image_url'] = (new Link())->getImageLink($product->link_rewrite[Context::getContext()->language->id], $img['id_image'], 'large_default');
+            $line['gtin'] = $line['product_ean13'] ?? $line['product_isbn'];
+
```
Wasn't there also mpn and brand in Prestashop?

to6ko12 replied:
MPN is added, but there is no column for brand in the products table. Should I add it?

TsankoBg replied:
👍  yes

TsankoBg commented on common/src/Module.php:
```diff
@@ -422,6 +450,22 @@ private function sendInvites($ps_shop_id) {
+            $img = $product->getCover($product->id);
+            $line['name'] = $line['product_name'];
+            $line['url'] = (new Link())->getProductLink($product);
+            $line['external_id'] = $line['product_id'];
```
There is no `external_id` in the API 
https://docs.webwinkelkeur.nl/api/invitations/add/#example-code

TsankoBg commented on common/src/Module.php:
```diff
@@ -422,6 +450,22 @@ private function sendInvites($ps_shop_id) {
         }
     }
 
+    private function getParsedOrderLines($order_lines) {
```
I would suggest renaming this to `getProducts`
```suggestion
    private function getProducts($order_lines) {
```

TsankoBg commented on common/src/Module.php:
```diff
@@ -442,13 +486,20 @@ public function getContent() {
+                $sync_url = _PS_BASE_URL_ . __PS_BASE_URI__ . sprintf('module/%s/sync', $this->getName());
+                $shop_id = Configuration::get($this->getConfigName('SHOP_ID'));
+                $this->sendSyncUrl($sync_url, $shop_id);
+            }
```
This is duplicated. Also you already do this check inside `sendSyncUrl`

TsankoBg commented on common/src/sync.php:
```diff
@@ -0,0 +1,110 @@
+use Product;
+use Symfony\Component\HttpKernel\Exception\HttpException;
+
+class sync extends ModuleFrontController {
```
This probably can be capitalized?

TsankoBg commented on common/src/sync.php:
```diff
@@ -0,0 +1,110 @@
+            throw new HttpException(404, sprintf('Could not find product with ID (%d)', $request_data['product_review']['product_id']));
+        }
+
+        if (!Configuration::get(strtoupper($this->module->name) . '_SYNC_PROD_REVIEWS')) {
```
Why don't you use `getConfigName`?

TsankoBg commented on common/src/sync.php:
```diff
@@ -0,0 +1,110 @@
+            throw new HttpException(403, 'Product review sync is disabled.');
+        }
+
+        $this->isAuthorized($request_data['webshop_id'], $request_data['api_key']);
```
I would move this check a bit higher, after `hasCredentialFields` and `credentialsEmpty`

TsankoBg commented on common/src/sync.php:
```diff
@@ -0,0 +1,110 @@
+            || !$product_review['reviewer']['name']
+            || !$product_review['rating']
+        ) {
+            throw new HttpException(404, 'Missing required content for product review');
```
`404` is a `not found` error message which is not appropriate in this case.
I would suggest removing this check. Our system does not allow product reviews with empty content.

TsankoBg commented on common/src/sync.php:
```diff
@@ -0,0 +1,110 @@
+        $product_comment_entity = new ProductComment();
+        $date_add = DateTime::createFromFormat('Y-m-d H:i:s', date('Y-m-d H:i:s'));
+        $product_comment_entity->setProductId($product_review['product_id'])
+            ->setCustomerId(0)
```
What happens when you create a pr with an empty CustomerId?
I think you should retrieve the customer with the product review email, and use this customer ID here. In case customer is not found, then use `0`

TsankoBg commented on common/src/sync.php:
```diff
@@ -0,0 +1,110 @@
+            ->setCustomerName($product_review['reviewer']['name'])
+            ->setGrade($product_review['rating'])
+            ->setValidate(1)
+            ->setDeleted($product_review['deleted'] ? 1 : 0)
```
Probably no need for `? 1 : 0`

TsankoBg commented on common/src/sync.php:
```diff
@@ -0,0 +1,110 @@
+        }
+        $entity_manager = $this->container->get('doctrine.orm.entity_manager');
+        $product_comment_entity = new ProductComment();
+        $date_add = DateTime::createFromFormat('Y-m-d H:i:s', date('Y-m-d H:i:s'));
```
Doesn't this always return the current datetime?

TsankoBg commented on common/src/sync.php:
```diff
@@ -0,0 +1,110 @@
+
+        $review_id = $product_comment_entity->getId();
+        if (!$review_id) {
+            $this->ajaxRender('There has been an error syncing the review in PrestaShop');
```
Here you should return an error, instead of just some text. 
It is a communication between 2 API endpoints, not 2 people.

TsankoBg commented on common/src/sync.php:
```diff
@@ -0,0 +1,110 @@
+        $entity_manager->flush();
+
+        $review_id = $product_comment_entity->getId();
+        if (!$review_id) {
```
What happens when the review gets deleted, does the review still exists? In magento2 I beleive we return `review_id => null`

to6ko12 replied:
The review still exists in Prestashop, so here it returns the review ID.

TsankoBg commented on common/src/sync.php:
```diff
@@ -0,0 +1,110 @@
+    private function isAuthorized(string $webshop_id, string $api_key): void {
+        $curr_webshop_id = Configuration::get(strtoupper($this->module->name) . '_SHOP_ID');
+        $curr_api_key = Configuration::get(strtoupper($this->module->name) . '_API_KEY');
+        if ($webshop_id == $curr_webshop_id && hash_equals($api_key, $curr_api_key)) {
```
Why don't you use `getConfigName`?

TsankoBg commented on common/templates/config_form.php:
```diff
@@ -141,3 +184,16 @@
+    }
+
+    return 'installed_and_enabled';
+}
```
Better use `!Module::isInstalled('productcomments')` , `!Module::isEnabled('productcomments')`  directly inside the ifs

to6ko12 committed:
Changes after PR
Also changed the logic, so when a product review is deleted, the review ID is returned, as reviews can be kept in Prestashop after deletion.

to6ko12 committed:
Get customer id
Correctly get customer id by email for product reviews

TsankoBg committed:
Call Tools::getValue('sync_prod_reviews') directly

TsankoBg commented on common/src/Sync.php:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    private function logReviewSync(string $review_id, bool $deleted = false) {
+        $deleted
+            ? \PrestaShopLogger::addLog(sprintf('Deleted product review with ID (%d)', $review_id))
+            : \PrestaShopLogger::addLog(sprintf('Saved product review with ID (%d)', $review_id));
```
No need to repeat this twice if the only difference is Saved/Deleted

```suggestion
\PrestaShopLogger::addLog(sprintf('%s product review with ID (%d)', $deleted ? 'Deleted' : 'Saved', $review_id));
```

TsankoBg commented on common/templates/config_form.php:
```diff
@@ -141,3 +187,21 @@
 <?php endif; ?>
+
+<?php
+function isReviewModuleInstalled(): string {
```
I think this is not used anymore?

TsankoBg commented on common/src/Sync.php:
```diff
@@ -0,0 +1,129 @@
+        if ($product_review['id']) {
+            $product_comment_repository = $entity_manager->getRepository(ProductComment::class);
+            $product_comment = $product_comment_repository->find($product_review['id']);
+            if ($product_review['deleted']) {
+                $product_comment->setDeleted(1);
+                $this->logReviewSync($product_review['id'], true);
+            } else {
+                $product_comment->setDeleted(0);
+                $this->logReviewSync($product_review['id']);
+            }
```
This is basically the same as 
```PHP
$product_comment->setDeleted($product_review['deleted']);
$this->logReviewSync($product_review['id'], $product_review['deleted']);
```

no need for the conditions

TsankoBg commented on common/src/Sync.php:
```diff
@@ -0,0 +1,129 @@
+            }
+            $entity_manager->flush();
+            $this->ajaxRender(json_encode(['review_id' => $product_review['id']], JSON_PARTIAL_OUTPUT_ON_ERROR));
+            die();
```
In case the existing review has been changed, e.g. changed text/rating/title, the review will not be updated.

TsankoBg reviewed (changes requested):
There are a lot of functions without type hints.

to6ko12 committed:
Add type hints + cleanup + edit reviews functionality

to6ko12 committed:
Merge remote-tracking branch 'origin/product_reviews' into product_reviews

to6ko12 committed:
Remove duplicated code

TsankoBg commented on common/src/Sync.php:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    private function logReviewSync(string $review_id, bool $deleted = false): void {
+        \PrestaShopLogger::addLog(sprintf('%s product review with ID (%d)', $deleted, $review_id));
```
```suggestion
        \PrestaShopLogger::addLog(sprintf('%s product review with ID (%d)', $deleted ? 'Deleted' : 'Saved', $review_id));
```

TsankoBg commented on common/src/Sync.php:
```diff
@@ -0,0 +1,114 @@
+        $customer = new Customer();
+        $customer->getByEmail($email);
+
+        return $customer->id;
```
This should be tested, probably does not return int when not found

to6ko12 committed:
Fix type hint

to6ko12 committed:
Update common/src/Sync.php

Co-authored-by: Tsanko Hadzhiev <tsankohadzhiev@gmail.com>

to6ko12 committed:
Change WWK logo + add warning in settings
